pubsublite/internal/test/msg_tracker.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package test
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MsgTracker is a helper for checking whether a set of messages make a full
+// round trip from publisher to subscriber.
+type MsgTracker struct {
+	msgMap map[string]bool
+	done   chan struct{}
+	mu     sync.Mutex
+}
+
+// NewMsgTracker creates a new message tracker.
+func NewMsgTracker() *MsgTracker {
+	return &MsgTracker{
+		msgMap: make(map[string]bool),
+		done:   make(chan struct{}, 1),
+	}
+}
+
+// Add a set of tracked messages.
+func (mt *MsgTracker) Add(msgs ...string) {
```

Comment:
IIUC, this type is designed to be used only once; that is, the user is not supposed to interleave calls to Add() and Remove(). Can we pass the full set of messages in NewMsgTracker() instead of having an Add() method?

Reply:
My longrunning integration tests loop and can reuse the same MsgTracker instance (to save having to add a mutex to its parent in order to replace it). Added a comment about correct usage of MsgTracker and pointed out the fact that Add and Remove are not supposed to be interleaved.